README.md

```diff
@@ -1,2 +1,182 @@
-# simple-worker
-Teleport Code Challenge
+
+
+# Code Challenge - Teleport
+
+Purpose of this project is to create a small service that is able to `Start`, `Stop`, get `Status` and stream `Output` of a Linux job. The complete package will consist of 3 main components:  
+1. Library to achieve above functionalities (used by server)  
+1. gRPC to achieve communication between clients and the server 
+1. CLI for clients to send requests to server to use above services
+
+
+## Library
+The library will provide functions to create, stop, query status and stream output of a job. It will use `os/exec` package to achieve most of its functionalities.
+
+The library will **not** store created jobs internally in memory. It will simply return the `Job` to the user and it is up to the user to manage them. This library will provide all the utilities needed to `Create`, `Stop`, `GetStatus`, and `Output` of a job.
+
+### Data structures
+```go
+type Job struct {
+	Cmd    *exec.Cmd
+	Pid    int
+
+	status JobStatus	// Status of this Job
+	*sync.Mutex			// to protect Status from simultaneous reads and writes
+	outStream Streamer	// used for creating streaming channel via Output()
+	doneC     chan struct{} // channel for user to block until job exits
+}
+
+type JobStatus {
+	Status string  // string representation of current status
+	Exited bool    // true if job has exited
+	ExitCode int   // exit code of the exited job
+	// extend as needed
+}
+
+type Streamer interface {
+	io.WriteCloser	// implements Writer and Closer
+	NewReader() (io.ReadCloser, error)
+}
+
+// streamer structure embeds os.File for implementing
+// simultaneous output streaming for multiple clients.
+type streamer struct {
+	*os.File // embeds os.File
+}
+
+```
+  Note that `streamer` structure embeds `os.File` and has many benefits:
+  1. Read() and Write() methods for free
+  2. Synchronization by kernel for free
+  3. Memory-error safe as data is stored in disk
+  4. Will be asigned to exec.Cmd.Stdout and exec.Cmd.Stderr fields and both outputs will go into a single file in chronological order.
+
+Say, if a client is requesting for a job that builds a project, logs can be very long and may take up a lof of space. The above approach is very handy as it uses disk space rather than in-memory buffers. And implementation will be simple as many things come for free.
```

Comment:
On another thought, a separate go routine can run in an infinite loop, iterating once in a while, and cleanup some of the old jobs. a TTL (Time To Live) kind of timer can be configured for each job and once it expires, it would trigger a function which pushes its UUID to a channel of cleanup go routine. When cleanup go routine sees some entries in that channel, it releases all the resources for that job.
I will not implement this for challenge but will add this info as a comment in the code.